tests/test_api_requests.py

```diff
@@ -325,3 +228,1149 @@ async def parse(self, response):
+            for method in (
+                "POST",
+                "PUT",
+                "DELETE",
+                "OPTIONS",
+                "TRACE",
+                "PATCH",
+                "HEAD",
+                "CONNECT",
+                "FOO",
+            )
+        ),
+        *(
+            (
+                request_method,
+                {"httpRequestMethod": meta_method},
+                {
+                    "httpResponseBody": True,
+                    "httpResponseHeaders": True,
+                    "httpRequestMethod": meta_method,
+                },
+                ["Use Request.method"],
+            )
+            for request_method, meta_method in (
+                ("GET", "GET"),
+                ("POST", "POST"),
+            )
+        ),
+        *(
+            (
+                request_method,
+                {"httpRequestMethod": meta_method},
+                {
+                    "httpResponseBody": True,
+                    "httpResponseHeaders": True,
+                    "httpRequestMethod": meta_method,
+                },
+                [
+                    "Use Request.method",
+                    "does not match the Zyte Data API httpRequestMethod",
+                ],
+            )
+            for request_method, meta_method in (
+                ("GET", "POST"),
+                ("PUT", "GET"),
+            )
+        ),
+        (
+            "POST",
+            {"browserHtml": True},
+            {
+                "browserHtml": True,
+                "httpResponseHeaders": True,
+            },
+            ["can only be set when the httpResponseBody parameter"],
+        ),
+        (
+            "POST",
+            {"screenshot": True},
+            {
+                "screenshot": True,
+            },
+            ["can only be set when the httpResponseBody parameter"],
+        ),
+    ],
+)
+def test_automap_method(method, meta, expected, warnings, caplog):
+    _test_automap({"method": method}, meta, expected, warnings, caplog)
+
+
+@pytest.mark.parametrize(
+    "headers,meta,expected,warnings",
+    [
+        # Base header mapping scenarios for a supported header.
+        (
+            {"Referer": "a"},
+            {},
+            {
+                "customHttpRequestHeaders": [
+                    {"name": "Referer", "value": "a"},
+                ],
+                "httpResponseBody": True,
+                "httpResponseHeaders": True,
+            },
+            [],
+        ),
+        (
+            {"Referer": "a"},
+            {"browserHtml": True},
+            {
+                "browserHtml": True,
+                "httpResponseHeaders": True,
+                "requestHeaders": {"referer": "a"},
+            },
+            [],
+        ),
+        (
+            {"Referer": "a"},
+            {"browserHtml": True, "httpResponseBody": True},
+            {
+                "browserHtml": True,
+                "customHttpRequestHeaders": [
```

Comment:
To clarify, wouldn't using `customHttpRequestHeaders` alongside `browserHtml=True` issue a warning?

Since the [doc](https://docs.zyte.com/zyte-api/openapi.html#operation/extract) mentions:
> `customHttpRequestHeaders`: This feature cannot be used with browserHtml, only with httpResponseBody

Reply:
I am assuming that Zyte API may at some point in the future allow combining browserHtml and httpResponseBody, and if we make that assumption, in this specific scenario, there should be no warning, because `httpResponseBody` is also set to `True`. Notice that in this scenario it is not the user passing `customHttpRequestHeaders`, that value comes from automated mapping.

I have no strong opinion on supporting combining browserHtml and httpResponseBody, though. But then we would have to decide which one to send when both are defined. And we may need to revisit what to do when none of the known current main outputs (httpResponseBody, browserHtml, screenshot) are `True`.

Reply:
> In any case, I see that there is no test covering the case where a user sets `browserHtml=True` and `customHttpRequestHeaders={…}`, but not `httpResponseBody`. I will extend tests accordingly.

[This is done now](https://github.com/scrapy-plugins/scrapy-zyte-api/pull/41/commits/1bad96bb9b5a3664f24a778e68e435004abb759e).